src/contracts/atlas/Escrow.sol

```diff
@@ -39,87 +39,105 @@ abstract contract Escrow is AtlETH {
+        withLockPhase(ExecutionPhase.PreOps)
+        returns (bytes memory)
     {
-        preOpsData = abi.encodeCall(IExecutionEnvironment.preOpsWrapper, userOp);
-        preOpsData = abi.encodePacked(preOpsData, lockBytes);
-        (success, preOpsData) = environment.call(preOpsData);
+        (bool success, bytes memory data) = ctx.executionEnvironment.call(
+            abi.encodePacked(
+                abi.encodeCall(IExecutionEnvironment.preOpsWrapper, userOp), ctx.setAndPack(ExecutionPhase.PreOps, true)
+            )
+        );
+
         if (success) {
-            preOpsData = abi.decode(preOpsData, (bytes));
+            if (dConfig.callConfig.needsPreOpsReturnData()) {
+                return abi.decode(data, (bytes));
+            } else {
+                return new bytes(0);
+            }
+        } else {
+            if (ctx.isSimulation) revert PreOpsSimFail();
+            revert PreOpsFail();
         }
     }
 
     /// @notice Executes the user operation logic defined in the Execution Environment.
+    /// @param ctx Metacall context data from the Context struct.
+    /// @param dConfig Configuration data for the DApp involved, containing execution parameters and settings.
     /// @param userOp UserOperation struct containing the user's transaction data.
-    /// @param environment Address of the execution environment where the user operation will be executed.
-    /// @param lockBytes Packed bytes form of the current escrow lock state, encoded from the EscrowKey struct.
-    /// @return success Boolean indicating whether the UserOperation was executed successfully.
     /// @return userData Data returned from executing the UserOperation, if the call was successful.
     function _executeUserOperation(
+        Context memory ctx,
+        DAppConfig calldata dConfig,
         UserOperation calldata userOp,
-        address environment,
-        bytes memory lockBytes
+        bytes memory returnData
     )
         internal
-        returns (bool success, bytes memory userData)
+        withLockPhase(ExecutionPhase.UserOperation)
+        returns (bytes memory)
     {
-        userData = abi.encodeCall(IExecutionEnvironment.userWrapper, userOp);
-        userData = abi.encodePacked(userData, lockBytes);
-
-        (success, userData) = environment.call{ value: userOp.value }(userData);
+        (bool success, bytes memory data) = ctx.executionEnvironment.call{ value: userOp.value }(
+            abi.encodePacked(
+                abi.encodeCall(IExecutionEnvironment.userWrapper, userOp),
+                ctx.setAndPack(ExecutionPhase.UserOperation, true)
+            )
+        );
 
         if (success) {
-            userData = abi.decode(userData, (bytes));
+            // Handle formatting of returnData
+            if (dConfig.callConfig.needsUserReturnData()) {
+                return abi.decode(data, (bytes));
+            } else {
+                return returnData;
```

Comment:
if needsUserReturnData is false, and needsPreOpsCall is true, then here we make the `returnData` of the preOpsCall available to all the solvers, which might be unintentional. should we return empty bytes here instead?

Reply:
https://github.com/FastLane-Labs/atlas/pull/227/commits/dd80b3a7018a846f95a8faa370b0f38f239bc398